src/libNode/Node.cpp

```diff
@@ -1782,55 +1783,68 @@ bool Node::ProcessTxnPacketFromLookup([[gnu::unused]] const bytes& message,
-      m_txnPacketBuffer.emplace(msg_hash, message2);
+      m_txnPacketBuffer.emplace(msg_hash, message);
       return true;
     }
   }
 
+  // for shard:
+  // 1. DS epoch: lookup dispatch in new DS epoch, buffer and dispatch when mb
+  // finish, so normally first DS epoch won't have txns
+  // 2. FB epoch: saying the 2nd epoch after ds epoch, lookup dispatch txn upon
+  // mb, they distribute right away until mb consensus started. During the mb
+  // consensus, all the packet received should be buffered
+  // for DS:
+  // 1. DS epoch: lookup dispatch in new DS epoch, distribute immediately until
+  // DSMB started, during DSMB and FB consensus, all packet should be buffered
+  // until the next epoch
+
   bool fromLookup = m_mediator.m_lookup->IsLookupNode(from) &&
                     from.GetPrintableIPAddress() != "127.0.0.1";
 
   bool properState =
       (m_mediator.m_ds->m_mode != DirectoryService::Mode::IDLE &&
        m_mediator.m_ds->m_state == DirectoryService::MICROBLOCK_SUBMISSION) ||
+
       (m_mediator.m_ds->m_mode != DirectoryService::Mode::IDLE &&
        m_mediator.m_node->m_myshardId == 0 && m_txn_distribute_window_open &&
        m_mediator.m_ds->m_state ==
            DirectoryService::FINALBLOCK_CONSENSUS_PREP) ||
+
       (m_mediator.m_ds->m_mode == DirectoryService::Mode::IDLE &&
        m_txn_distribute_window_open &&
        (m_state == MICROBLOCK_CONSENSUS_PREP ||
-        m_state == MICROBLOCK_CONSENSUS));
+        m_state == MICROBLOCK_CONSENSUS || m_state == WAITING_FINALBLOCK));
 
-  if (fromLookup || !properState) {
+  if (!properState) {
     if ((epochNumber + (fromLookup ? 0 : 1)) < m_mediator.m_currentEpochNum) {
       LOG_GENERAL(WARNING, "Txn packet from older epoch, discard");
       return false;
     }
     lock_guard<mutex> g(m_mutexTxnPacketBuffer);
-    LOG_GENERAL(INFO, string(fromLookup ? "Received txn packet from lookup"
-                                        : "Received not in the proper state") +
-                          ", store txn packet to buffer");
+    LOG_GENERAL(INFO,
+                "Received not in the proper state, store txn packet to buffer");
     if (fromLookup) {
       LOG_STATE("[TXNPKTPROC]["
                 << std::setw(15) << std::left
                 << m_mediator.m_selfPeer.GetPrintableIPAddress() << "]["
                 << m_mediator.m_currentEpochNum << "][" << shardId << "]["
-                << string(lookupPubKey).substr(0, 6) << "][" << message2.size()
+                << string(lookupPubKey).substr(0, 6) << "][" << message.size()
                 << "] RECVFROMLOOKUP");
     }
     SHA2<HashType::HASH_VARIANT_256> sha256;
-    sha256.Update(message2);  // message hash
+    sha256.Update(message);  // message hash
     bytes msg_hash = sha256.Finalize();
-    m_txnPacketBuffer.emplace(msg_hash, message2);
+    m_txnPacketBuffer.emplace(msg_hash, message);
   } else {
+    // not from lookup and in proper state
     LOG_GENERAL(INFO,
```

Comment:
This message may not be always true. since we do process messages received directly from lookup now due to change at line Line 1822.